apps/provisioning/pkg/auth/token_access_checker.go

```diff
@@ -0,0 +1,62 @@
+package auth
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	authlib "github.com/grafana/authlib/types"
+	"github.com/grafana/grafana/pkg/apimachinery/identity"
+)
+
+// tokenAccessChecker implements AccessChecker using access tokens from context.
+type tokenAccessChecker struct {
+	inner authlib.AccessChecker
+}
+
+// NewTokenAccessChecker creates an AccessChecker that gets identity from access tokens
+// via AuthInfoFrom(ctx). Role-based fallback is not supported.
+func NewTokenAccessChecker(inner authlib.AccessChecker) AccessChecker {
+	return &tokenAccessChecker{inner: inner}
+}
+
+// WithFallbackRole returns the same checker since fallback is not supported.
+func (c *tokenAccessChecker) WithFallbackRole(_ identity.RoleType) AccessChecker {
+	return c
+}
+
+// Check performs an access check using AuthInfo from context.
+// Returns nil if access is allowed, or an appropriate API error if denied.
+func (c *tokenAccessChecker) Check(ctx context.Context, req authlib.CheckRequest, folder string) error {
+	// Get identity from access token in context
+	id, ok := authlib.AuthInfoFrom(ctx)
+	if !ok {
+		return apierrors.NewUnauthorized("no auth info in context")
+	}
+
+	// AccessPolicy identities are trusted internal callers
+	if authlib.IsIdentityType(id.GetIdentityType(), authlib.TypeAccessPolicy) {
```

Comment:
Is this always the case? I don't think so when this code is running in the MT Service.

In ST service, we have abused TypeAccessPolicy in lieu of Service Identity which was supposed to be used in-process only.

When we pepper our code with such checks, its hard to reason whether the identity type in this case is coming in from the wire, or from within process. If its coming from the wire, then we need to ensure adequate permissions here or somewhere down the chain).

On another note, we should extend the permissions [here](https://github.com/grafana/grafana/blob/f3d4181cf27be01384654af6645d14e87e5afd42/pkg/apimachinery/identity/context.go#L129-L153) and actually verify that these permissions hold, instead of trusting them blindly.

Reply:
yea, I changed the behavior. Well spotted.

Reply:
> On another note, we should extend the permissions [here](https://github.com/grafana/grafana/blob/f3d4181cf27be01384654af6645d14e87e5afd42/pkg/apimachinery/identity/context.go#L129-L153) and actually verify that these permissions hold, instead of trusting them blindly.

Aren't those already validated if we call `access.Check` for the context? How would that work? @charandas